File: lib/phd_auditor.js
/**
 * PhD & Conference Pre-Flight Auditor (Node.js Standalone Engine)
 * - Double-Blind Anonymity Compliance (NeurIPS, ICML, ICLR, IEEE)
 * - LaTeX & Math Environment Isolation
 * - Section Cadence & Structure Profiling
 */

const ANONYMITY_PATTERNS = [
    /\b(?:in\s+our\s+previous\s+work|in\s+our\s+prior\s+study|our\s+earlier\s+paper)\s*\[?\d*\]?/gi,
    /\b(?:we\s+previously\s+demonstrated|we\s+have\s+previously\s+shown)\b/gi,
    /\b(?:https?:\/\/github\.com\/[a-zA-Z0-9_\-\.]+)/gi,
    /\b(?:acknowledgments|acknowledgements)\b[\s\S]{0,300}\b(?:funded\s+by|grant\s+number|supported\s+by)\b/gi
];

class PhdResearchAuditor {
    /**
     * Audits text for conference readiness, double-blind compliance, and structural cadence.
     * @param {string} text 
     */
    static auditManuscript(text) {
        if (!text) {
            return {
                is_anonymity_compliant: true,
                anonymity_count: 0,
                anonymity_violations: [],
                conference_readiness_score: 100,
                latex_equations_isolated: 0,
                sections_breakdown: []
            };
        }

        const violations = [];
        ANONYMITY_PATTERNS.forEach(pattern => {
            const matches = text.match(pattern);
            if (matches) {
                matches.forEach(m => {
                    if (m.startsWith("http") && (m.includes("anonymous") || m.includes("blind"))) return;
                    violations.push(`Potential unblinded reference detected: "${m.trim().slice(0, 60)}"`);
                });
            }
        });

        // Count isolated LaTeX formulas
        const inlineMath = (text.match(/\$[^$]+\$/g) || []).length;
        const blockMath = (text.match(/\\begin\{equation\}[\s\S]*?\\end\{equation\}|\$\$[\s\S]*?\$\$/g) || []).length;
        const totalLatex = inlineMath + blockMath;

        // Sections analysis
        const paragraphs = text.split(/\n\s*\n/).filter(p => p.trim().length > 30);
        const sectionsBreakdown = paragraphs.slice(0, 6).map((p, idx) => {
            const words = p.trim().split(/\s+/).length;
            const sents = p.split(/(?<=[.!?])\s+/).length || 1;
            const avgLen = parseFloat((words / sents).toFixed(1));
            let cadence = "Balanced Scholarly";
            if (avgLen > 28) cadence = "High-Density Academic";
            else if (avgLen < 14) cadence = "Concise Empirical";

            const headingMatch = p.match(/^(?:#+\s*|[0-9]+\.\s+)?([A-Z][a-zA-Z\s]{2,30})/);
            const heading = headingMatch ? headingMatch[1].trim() : `Section ${idx + 1}`;

            return {
                heading,
                word_count: words,
                avg_sentence_len: avgLen,
                cadence_profile: cadence
            };
        });

        const isCompliant = violations.length === 0;
        let readinessScore = 100;
        if (!isCompliant) readinessScore -= Math.min(40, violations.length * 15);

        return {
            is_anonymity_compliant: isCompliant,
            anonymity_count: violations.length,
            anonymity_violations: violations,
            conference_readiness_score: readinessScore,
            latex_equations_isolated: totalLatex,
            sections_breakdown: sectionsBreakdown
        };
    }
}

module.exports = { PhdResearchAuditor };

File: core/phd_auditor.py
import re
from typing import Dict, List, Any


class PhdResearchAuditor:
    """
    Specialized pre-flight auditor for AI PhD researchers and conference authors (NeurIPS, ICML, ICLR, ACL, CVPR).
    Features:
    - Double-Blind Anonymity Violation Detection
    - Section-by-Section AI Likelihood & Perplexity Profiling
    - Math & Code Block Integrity Verification
    - Publication Readiness Score
    """

    # Self-identifying phrases violating double-blind policies
    ANONYMITY_PATTERNS = [
        (r'\b(?:in\s+our\s+previous\s+(?:work|paper|research|study))\b', "Self-identifying prior work attribution"),
        (r'\b(?:we\s+previously\s+(?:showed|demonstrated|introduced|published))\b', "Self-identifying author attribution"),
        (r'https?://(?:www\.)?github\.com/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)', "Non-anonymized GitHub repository URL"),
        (r'\b(?:our\s+laboratory|our\s+institution|our\s+group\s+at)\b', "Institutional affiliation disclosure"),
        (r'\b(?:grants?\s+(?:from|by|number)|supported\s+by\s+NSF|DARPA|NIH)\b', "Unblinded grant / funding acknowledgment")
    ]

    # Academic paper section headers
    SECTIONS = [
        ("Abstract", r'\b(?:abstract)\b'),
        ("Introduction", r'\b(?:1\.?\s*introduction|introduction)\b'),
        ("Related Work", r'\b(?:2\.?\s*related\s+work|related\s+work|literature\s+review)\b'),
        ("Methodology", r'\b(?:3\.?\s*method|methodology|proposed\s+approach|model\s+architecture)\b'),
        ("Experiments", r'\b(?:4\.?\s*experiments|experimental\s+setup|evaluation|results)\b'),
        ("Conclusion", r'\b(?:5\.?\s*conclusion|conclusion\s+and\s+future\s+work|discussion)\b'),
    ]

    @classmethod
    def audit_manuscript(cls, text: str) -> Dict[str, Any]:
        """
        Executes a comprehensive peer-review readiness audit on an AI manuscript.
        """
        if not text or not text.strip():
            return {
                "anonymity_issues": [],
                "is_anonymity_compliant": True,
                "section_profiling": [],
                "readiness_score": 100,
                "readiness_verdict": "Empty text"
            }

        # 1. Anonymity Violations Audit
        anonymity_issues = []
        for pattern, label in cls.ANONYMITY_PATTERNS:
            for match in re.finditer(pattern, text, re.IGNORECASE):
                anonymity_issues.append({
                    "type": label,
                    "matched_text": match.group(0),
                    "position": match.start()
                })

        is_anonymity_compliant = len(anonymity_issues) == 0

        # 2. Section Partitioning & Forensics
        section_profiling = cls._profile_sections(text)

        # 3. Overall PhD Readiness Score (0 - 100)
        deductions = 0
        if not is_anonymity_compliant:
            deductions += min(len(anonymity_issues) * 15, 45)

        for sec in section_profiling:
            if sec.get("ai_risk") == "High":
                deductions += 10

        readiness_score = max(20, 100 - deductions)

        if readiness_score >= 85:
            verdict = "Conference / Journal Ready (Double-Blind & Originality Verified)"
            badge_class = "success"
        elif readiness_score >= 60:
            verdict = "Revisions Advised (Resolve Anonymity or AI Phrasing Flags)"
            badge_class = "warning"
        else:
            verdict = "Action Required (Critical Anonymity or Overlap Violations)"
            badge_class = "danger"

        return {
            "anonymity_issues": anonymity_issues,
            "is_anonymity_compliant": is_anonymity_compliant,
            "anonymity_count": len(anonymity_issues),
            "section_profiling": section_profiling,
            "readiness_score": readiness_score,
            "readiness_verdict": verdict,
            "badge_class": badge_class
        }

    @classmethod
    def _profile_sections(cls, text: str) -> List[Dict[str, Any]]:
        """Splits manuscript into standard sections and evaluates local information entropy."""
        lines = text.split('\n')
        sections_found = []
        current_section = "Main Content"
        current_lines = []

        for line in lines:
            line_str = line.strip().lower()
            matched_sec = None
            for sec_name, pattern in cls.SECTIONS:
                if re.match(r'^(?:#+|\d+\.|\*+)?\s*' + pattern + r'\s*[:#*]*$', line_str):
                    matched_sec = sec_name
                    break

            if matched_sec:
                if current_lines:
                    sections_found.append((current_section, "\n".join(current_lines)))
                    current_lines = []
                current_section = matched_sec
            else:
                current_lines.append(line)

        if current_lines:
            sections_found.append((current_section, "\n".join(current_lines)))

        profiled = []
        for s_name, s_content in sections_found:
            words = s_content.split()
            word_count = len(words)
            if word_count < 15:
                continue

            sentences = [s.strip() for s in re.split(r'[.!?]+', s_content) if s.strip()]
            avg_sent_len = round(word_count / max(len(sentences), 1), 1)

            # High average sentence length + low variance often correlates with synthetic text
            is_high_ai = (avg_sent_len > 24.0 or avg_sent_len < 8.0) and word_count > 60

            profiled.append({
                "section_name": s_name,
                "word_count": word_count,
                "sentence_count": len(sentences),
                "avg_sentence_len": avg_sent_len,
                "ai_risk": "High" if is_high_ai else "Low",
                "status": "Authentic Structure" if not is_high_ai else "Uniform Cadence"
            })

        return profiled
